### PR TITLE
state store: fix logic for evaluating job status

### PR DESCRIPTION
**Pull request by mismithhisler**
### Description

The state store persists the correct job version for job summaries, but not for specific versions.  This fix attempts to simplify the logic around evaluating a job's status, and then make's sure it is persisted.

### Testing & Reproduction steps

See GH issue for replication steps

### Links

Fixes [GH #24957](https://github.com/hashicorp/nomad/issues/24957)

### Contributor Checklist
- [x] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [x] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [x] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository.

### CODE REVIEW

**Comment by mismithhisler** [comment]
Currently failing some unit tests but I wanted to get a draft up for basic review of the logic change

File: nomad/state/state_store_test.go [hunk 1]
```diff
@@ -7678,232 +7678,100 @@ func TestStateStore_SetJobStatus(t *testing.T) {
 	}
 }
 
-func TestStateStore_GetJobStatus_NoEvalsOrAllocs(t *testing.T) {
+func TestStateStore_GetJobStatus_new(t *testing.T) {
```
Comment (mismithhisler) [inline nomad/state/state_store_test.go:7681]: Need to fix the name of this test.

File: nomad/state/state_store_test.go [hunk 2]
```diff
@@ -7678,232 +7678,100 @@ func TestStateStore_SetJobStatus(t *testing.T) {
...
+			exp:      structs.JobStatusDead,
+		},
+		{
+			name:     "previous job version had allocs",
+			hasAlloc: true,
+			jobSetup: func(j *structs.Job) {
+				j.Version += 1
+			},
+			exp: structs.JobStatusPending,
+		},
```
Comment (mismithhisler) [inline nomad/state/state_store_test.go:7744]: Need to add more test case for scenario involving pending evaluations.

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -3851,11 +3852,6 @@ func (s *StateStore) EvalsByJob(ws memdb.WatchSet, namespace, jobID string) ([]*
 
 		e := raw.(*structs.Evaluation)
 
-		// Filter non-exact matches
-		if e.JobID != jobID {
```
Comment (mismithhisler) [inline nomad/state/state_store.go:3855]: This isn't necessary anymore because periodic jobs do not get an evaluation.

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -5532,6 +5527,12 @@ func (s *StateStore) setJobStatus(index uint64, txn *txn,
 	if err := s.setJobSummary(txn, updated, index, oldStatus, newStatus); err != nil {
 		return fmt.Errorf("job summary update failed %w", err)
 	}
+
+	// Update the job version details
+	if err := s.upsertJobVersion(index, updated, txn); err != nil {
```
Comment (mismithhisler) [inline nomad/state/state_store.go:5532]: This feels odd to call here, but it's necessary to also update a job's version status anytime we are updating the job summary, or they can get out of sync.

File: nomad/state/state_store_test.go [hunk 5]
```diff
@@ -5084,6 +5087,7 @@ func TestStateStore_DeleteEval_Eval(t *testing.T) {
 	require.Equal(t, uint64(1002), evalsIndex)
 }
 
+// This tests the evalDelete boolean by deleting a Pending eval and Pending Alloc.
```
Comment (mismithhisler) [inline nomad/state/state_store_test.go:5090]: This job tests the `evalDelete` boolean, but it's an odd test because it forcefully deletes a non-terminal eval and non-terminal alloc, and then makes sure we set the job to "dead".

File: nomad/fsm.go [hunk 6]
```diff
@@ -948,20 +948,21 @@ func (n *nomadFSM) applyAllocClientUpdate(msgType structs.MessageType, buf []byt
...
 	}
 
-	// Update all the client allocations
-	if err := n.state.UpdateAllocsFromClient(msgType, index, req.Alloc); err != nil {
-		n.logger.Error("UpdateAllocFromClient failed", "error", err)
-		return err
-	}
-
-	// Update any evals
+	// Update any evals first.  During a reschedule, we don't want to mark the job dead, which
```
Comment (mismithhisler) [inline nomad/fsm.go:951]: This is probably not necessary anymore with the `isReschedulable` function check.

File: nomad/state/state_store.go [hunk 7]
```diff
@@ -5633,8 +5641,11 @@ func (s *StateStore) getJobStatus(txn *txn, job *structs.Job, evalDelete bool) (
 	for raw := evals.Next(); raw != nil; raw = evals.Next() {
 		e := raw.(*structs.Evaluation)
 
-		// Filter non-exact matches
-		if e.JobID != job.ID {
+		// Handles restarting stopped jobs and rescheduled allocs, or else they
+		// are briefly marked dead. We don't always want to skip these evaluations,
+		// like in the case of rescheduled or stopped jobs, but we handle both
+		// those cases in elsewhere in this function.
+		if e.JobModifyIndex < job.ModifyIndex {
```
Comment (mismithhisler) [inline nomad/state/state_store.go:5648]: Per this comment, we don't always want to filter out these evals, as they may still apply.  This is really a tradeoff between adding more logic to this function, or accepting we will set job statuses incorrectly for brief periods of time.

File: .changelog/24974.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+state store: fix for setting correct job status in various scenarios
```
Comment (tgross) [inline .changelog/24974.txt:2]: We should probably try to summarize what those scenarios are as best we can here.

File: nomad/state/state_store_test.go [hunk 9]
```diff
@@ -7678,232 +7685,228 @@ func TestStateStore_SetJobStatus(t *testing.T) {
...
+		},
+		{
+			name: "no allocs",
+			setup: func(txn *txn) (*structs.Job, error) {
+				return mock.Job(), nil
+			},
+			exp: structs.JobStatusPending,
+		},
+		{
+			name: "current job has running alloc",
```
Comment (tgross) [inline nomad/state/state_store_test.go:7760]:
```suggestion
			name: "current job has pending alloc",
```

File: nomad/state/state_store_test.go [hunk 10]
```diff
@@ -7678,232 +7685,228 @@ func TestStateStore_SetJobStatus(t *testing.T) {
...
-	}
+			txn := state.db.WriteTxn(0)
 
-	// Stop the job
-	job.Stop = true
-	status, err = state.getJobStatus(txn, job, true)
-	if err != nil {
-		t.Fatalf("getJobStatus() failed: %v", err)
-	}
+			job, _ := tc.setup(txn)
```
Comment (tgross) [inline nomad/state/state_store_test.go:7904]: Any error we get from the setup function is a setup invariant, and we're throwing it away. To make that more useful, we can pass the `testing.T` in and then call `must.NoError(t, err)` inside the setup function instead of returning errors from it:

```suggestion
			job := tc.setup(t, txn)
```

File: nomad/state/state_store_test.go [hunk 11]
```diff
@@ -7678,232 +7685,228 @@ func TestStateStore_SetJobStatus(t *testing.T) {
...
 
-	if expected := structs.JobStatusDead; status != expected {
-		t.Fatalf("getJobStatus() returned %v; expected %v", status, expected)
+			status, err := state.getJobStatus(txn, job, false)
+			require.NoError(t, err)
+			require.Equal(t, tc.exp, status)
```
Comment (tgross) [inline nomad/state/state_store_test.go:7908]:
```suggestion
			must.NoError(t, err)
			must.Eq(t, tc.exp, status)
```

File: nomad/state/state_store.go [hunk 12]
```diff
@@ -7348,6 +7363,30 @@ func (s *StateStore) ScalingPoliciesByIDPrefix(ws memdb.WatchSet, namespace stri
...
+		for j := len(a.RescheduleTracker.Events) - 1; j >= 0; j-- {
+			lastAttempt := a.RescheduleTracker.Events[j].RescheduleTime
+			timeDiff := currTime.UTC().UnixNano() - lastAttempt
+			if timeDiff < interval.Nanoseconds() {
+				attempted += 1
+			}
+		}
+	}
+	return attempted < availableAttempts
+}
```
Comment (tgross) [inline nomad/state/state_store.go:7388]: There's a method `(*RescheduleTracker).RescheduleEligible` we should be using here so that we're sharing identical logic between the state store and scheduler.

File: nomad/state/state_store_test.go [hunk 13]
```diff
@@ -7678,232 +7685,228 @@ func TestStateStore_SetJobStatus(t *testing.T) {
...
+		},
 	}
 
-	txn := state.db.ReadTxn()
-	status, err := state.getJobStatus(txn, job, true)
-	if err != nil {
-		t.Fatalf("getJobStatus() failed: %v", err)
-	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
```
Comment (tgross) [inline nomad/state/state_store_test.go:7899]: We can probably run this inner loop in parallel. Doesn't look like there's anything shared between tests.

File: nomad/state/state_store.go (hunk 4 above)
Comment (tgross) [reply nomad/state/state_store.go:5534]: Let's improve the comment here to point that out.

File: nomad/state/state_store.go (hunk 7 above)
Comment (tgross) [reply nomad/state/state_store.go:5648]: I'm not sure I understand this filter step or the associated comment. This ignores all evals except for the first one for a given version of the job (because that's the only time the `eval.JobModifyIndex` will match the `job.ModifyIndex`). How are those evals getting handled?

Also, why is it safe to remove the exact ID match now?

**Review by tgross — COMMENTED** [review]
This is looking like the right direction, but I've left some comments and questions. The new test looks great!

File: nomad/state/state_store.go (hunk 7 above)
Comment (mismithhisler) [reply nomad/state/state_store.go:5648]: I'm not in love with this logic.  _However_, there is an odd edge case at the moment where jobs that should be marked pending will be marked as dead for a brief period of time.  The best example I can give is after starting a previously stopped job.  We [upsert the job in a transaction](https://github.com/hashicorp/nomad/blob/main/nomad/fsm.go#L620), and then [upsert the eval](https://github.com/hashicorp/nomad/blob/main/nomad/fsm.go#L722) in a separate transaction.  So after the initial job upsert (but before eval upsert) the `getJobStatus()` will see that the job has all terminal evals, and mark it dead.

> Also, why is it safe to remove the exact ID match now?

This appears to be remnants of when we used to create evaluations for periodic jobs, and when we were determining the parent job's status, we had to filter out the child job's evals.  We don't create evals for periodic jobs anymore, so I removed this.

Comment (tgross) [reply nomad/state/state_store.go:5648]: > We upsert the job in a transaction, and then upsert the eval in a separate transaction. 

_Curses!_ That bit of code has been such a thorn!

**Review by tgross — APPROVED** [review]
LGTM! Nice work on this!
